pkg/metadata/sandbox.go

```diff
@@ -150,3 +164,115 @@ func (s *sandboxStore) List() ([]*SandboxMetadata, error) {
 	return s.store.Delete(sandboxID)
 }
+
+type SandboxNetNS struct {
+	sync.Mutex
+	NS       ns.NetNS
+	Symlink  *os.File
+	Closed   bool
+	Restored bool
+}
+
+func (ns *SandboxNetNS) symlinkCreate(name string) error {
+	b := make([]byte, 4)
+	_, randErr := rand.Reader.Read(b)
+	if randErr != nil {
+		return fmt.Errorf("failed to read random number: %v", randErr)
+	}
+	nsName := fmt.Sprintf("%s-%x", name, b)
+	symlinkPath := filepath.Join(nsRunDir, nsName)
+	if err := os.Symlink(ns.NS.Path(), symlinkPath); err != nil {
+		return fmt.Errorf("failed to symlink %q and %q: %v", ns.NS.Path(), symlinkPath, err)
+	}
+	fd, err := os.Open(symlinkPath)
+	if err != nil {
+		if removeErr := os.RemoveAll(symlinkPath); removeErr != nil {
+			return fmt.Errorf("failed to remove file %q: %v", symlinkPath, removeErr)
+		}
+		return fmt.Errorf("failed to open file %q: %v", symlinkPath, err)
+	}
+	ns.Symlink = fd
+	return nil
+}
+
+func (ns *SandboxNetNS) symlinkRemove() error {
+	if err := ns.Symlink.Close(); err != nil {
+		return fmt.Errorf("failed to close symlink: %v", err)
+	}
+	return os.RemoveAll(ns.Symlink.Name())
+}
+
+// NetNsPath gets the netns symlink stored in the sandbox meta data.
+func (s *SandboxMetadata) NetNSSymlink(createSandbox bool) string {
+	if s.NetNS == nil {
+		if createSandbox {
+			return ""
+		} else {
+			return getNetworkNamespace(s.SandboxPID)
+		}
+	}
+	return s.NetNS.Symlink.Name()
+}
+
+// NetNsPath returns the path to the network namespace of the container.
+func (s *SandboxMetadata) NetNsPath() string {
+	if s.NetNS == nil {
+		return getNetworkNamespace(s.SandboxPID)
+	}
+	return s.NetNS.NS.Path()
+}
+
+// NetNsCreate creates netns and symlink it with a file.
+func (s *SandboxMetadata) NetNSCreate() error {
+	if s.NetNS != nil {
+		return fmt.Errorf("net NS already created")
+	}
+	netNS, err := ns.NewNS()
+	if err != nil {
+		return fmt.Errorf("failed to create new net NS: %v", err)
+	}
+	s.NetNS = &SandboxNetNS{
+		NS:     netNS,
+		Closed: false,
+	}
+	if err := s.NetNS.symlinkCreate(s.Name); err != nil {
+		glog.Errorf("Could not create nentns symlink %v", err)
+		if closeErr := s.NetNS.NS.Close(); closeErr != nil {
+			return fmt.Errorf("net NS can not be close: %v", closeErr)
+		}
+		return fmt.Errorf("failed to create symlink for sandbox %q: %v", s.Name, err)
+	}
+	return nil
+}
+
+// NetNsRemove removes the netns store in sandbox meta data.
+func (s *SandboxMetadata) NetNSRemove() error {
+	glog.Error("remove bei diaoyongle")
+
+	if s.NetNS == nil {
+		glog.Error("no networking namespace to be removed.")
+		return nil
+	}
+	s.NetNS.Lock()
+	defer s.NetNS.Unlock()
+	if s.NetNS.Closed {
+		// netNsRemove() can be called multiple times without returning an error.
+		return nil
+	}
+	if err := s.NetNS.symlinkRemove(); err != nil {
+		return fmt.Errorf("failed to remove symlink: %v", err)
+	}
+	if err := s.NetNS.NS.Close(); err != nil {
+		return fmt.Errorf("failed to close net NS: %v", err)
+	}
+	if s.NetNS.Restored {
```

Comment:
This code is not useful because you didn't add the `netNsGet` function.

We don't need the restore logic for now, because we don't support restore after `cri-containerd` restart.
We can remove this code, and add a TODOs here:
1) We should handle restore logic with `netNsGet`.
2) We should get read of the forcibly cleanup on restore after https://github.com/containernetworking/cni/pull/342 is fixed and link the original issue https://github.com/kubernetes-incubator/cri-o/pull/276.